### PR TITLE
Provisional Python 2 and Python 3 compatibility 

### PR DESCRIPTION
**Pull request by tockards**
This adds Python 2 and Python 3 compatibility.

Putting the PR through for comments, suggestions and requests for changes due to the nature of the PR. 

All the test are passing. The tests that test unicode compatibility have been skipped in Python 3 for obvious reasons. 

## Notes

- The `ensure_native_str` was adapted from https://github.com/ska-sa/katcp-python/blob/4240fb45ec8cb73c4b60490e6e43ffe97700f567/katcp/compat.py#L17 
It no longer ensures native string and does the ascii replacement. The name has to change and is open for suggestions. 

- ~~@ajoubertza Certain compatibility code is duplicated, should I add a compat.py file similar to katcp-python library and add the code to those files?~~

- `itervalues` and `list` has been used. There are other methods and ways of achieving the same result and both ways impact performance. Input and comments on performance and which is better to use from those who use and wrote the library would be helpful.

- ~~The coverage has also been disabled due to the tests using a multi process module to run the tests and this breaks the coverage database. This might be due to the new image and a later version of nose and coverage and it is  something that needs to be looked into and outside of the scope of python 2 and python 3 compatibility.~~

### CODE REVIEW

File: tango_simlib/tests/test_sim_xmi_parser.py [hunk 1]
```diff
@@ -310,6 +311,7 @@ def test_attribute_list(self):
             "Actual tango device attribute list differs from expected " "list!",
         )
 
+    @unittest.skipIf((not future.utils.PY2), "Only for Python 2")
```
Comment (lanceWilliams) [inline tango_simlib/tests/test_sim_xmi_parser.py]: Don't we need to attempt these for Python3 as well? (assuming that is how this decorator works)

File: tango_simlib/tests/test_dish.py [hunk 2]
```diff
@@ -108,14 +111,14 @@ def setUpClassWithCleanup(cls):
...
-        cls.model = cls.models.values()[0]
+        cls.model = list(itervalues(cls.models))[0]
 
     def setUp(self):
         super(test_DishElementMaster, self).setUp()
         self.addCleanup(self._reset_model_defaults)
 
     def _reset_model_defaults(self):
-        for quantity in self.model.sim_quantities.values():
+        for quantity in list(itervalues(self.model.sim_quantities)):
```
Comment (lanceWilliams) [inline tango_simlib/tests/test_dish.py]: Isn't safe to just use the .values() in cases like this? In PY2 we would get the list and in PY3 we would get the iterator?

File: tango_simlib/tests/test_tango_sim_generator.py [hunk 3]
```diff
@@ -259,7 +260,7 @@ def test_device_attribute_list(self):
         not_added_attr_names = not_added_attr.value
 
         expected_attributes = []
-        for attr_prop in self.sim_file_parser._device_attributes.values():
+        for attr_prop in list(self.sim_file_parser._device_attributes.values()):
```
Comment (lanceWilliams) [inline tango_simlib/tests/test_tango_sim_generator.py]: Maybe no list casting needed here either? The tuts on this seems to say that it only be an issue if the 'self.sim_file_parser._device_attributes' dict was being modified in the loop somewhere?

File: tango_simlib/tests/test_dish.py (hunk 2 above)
Comment (tockards) [reply tango_simlib/tests/test_dish.py]: This was a defensive change to get the tests to work. following https://python-future.org/compatible_idioms.html#dictionaries

For the iterations I should remove the casting. Thanks.

File: tango_simlib/tests/test_sim_xmi_parser.py (hunk 1 above)
Comment (tockards) [reply tango_simlib/tests/test_sim_xmi_parser.py]: @lanceWilliams  not sure we need input from @kmadisa , @ajoubertza, @samueltwum1  
the description contains a unicode character that is replaced with a `?` and this is present in the `expected_pressure_attr_info` https://github.com/ska-sa/tango-simlib/blob/9b9ccb7df9477d92fe3b8d1ae2a5b12c79fe2ab0/tango_simlib/tests/test_sim_xmi_parser.py#L120
When this test is executed in Python3 the correct unicode value is returned and therefor I skipped the tests and the functionality is different in Python2 and Python3 . 

Perhaps the test should not be skipped and the ascii encoded value should be returned and not unicode in Python3 also?

File: Jenkinsfile [hunk 4]
```diff
@@ -78,7 +84,6 @@ pipeline {
                         autoUpdateHealth: true,
                         autoUpdateStability: true,
                         zoomCoverageChart: true,
-                        // TODO: The reason this is commented out is because tango-simlib test coverage is currently at 70% instead of minimum 80%.
```
Comment (mmphego) [inline Jenkinsfile]: Why was this deleted?

File: Jenkinsfile [hunk 5]
```diff
@@ -54,15 +54,21 @@ pipeline {
...
                     }
                 }
 
                 stage ('py36') {
                     steps {
-                         echo "Running nosetests on Python 3.6"
-                         sh 'python3 -m pip install -U .'
-                         sh "python3 setup.py nosetests --with-xunitmp --xunitmp-file=nosetests_py36.xml"
+                        echo "Running nosetests on Python 3.6"
+                        sh 'python3 -m pip install katcp'
```
Comment (mmphego) [inline Jenkinsfile]: We do not need to install katcp as it's baked into the image.

File: tango_simlib/compat.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
+# katcp.py
```
Comment (mmphego) [inline tango_simlib/compat.py]: Not sure what this line means? Is it supposed to be here - A bit misleading if you ask me.

File: tango_simlib/compat.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    def ensure_native_str(value):
+        """Coerce unicode string or bytes to native string type (UTF-8 encoding)."""
+        if isinstance(value, str):
+            return value
+        elif isinstance(value, unicode): # noqa
+            return value.encode("ascii", "replace")
+        else:
+            raise TypeError("Invalid type for string conversion: {}".format(type(value)))
+
+
```
Comment (mmphego) [inline tango_simlib/compat.py]: Too many whitespaces

File: tango_simlib/compat.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+import future
+
+if future.utils.PY2:
+
+    def ensure_native_str(value):
+        """Coerce unicode string or bytes to native string type (UTF-8 encoding)."""
+        if isinstance(value, str):
+            return value
+        elif isinstance(value, unicode): # noqa
```
Comment (mmphego) [inline tango_simlib/compat.py]: Perhaps run `black .` to fix all the styling related issues.

File: tango_simlib/model.py [hunk 9]
```diff
@@ -244,10 +246,10 @@ def setup_sim_quantities(self):
                 # i.e. All optional parameters not provided in the SimDD
                 attr_props = dict(
                     (param_key, param_val)
-                    for param_key, param_val in attr_props.iteritems()
+                    for param_key, param_val in iteritems(attr_props)
                     if param_val
                 )
-                model_attr_props = dict(model_attr_props.items() + attr_props.items())
+                model_attr_props.update(attr_props)
```
Comment (mmphego) [inline tango_simlib/model.py]: Nice :+1:

File: tango_simlib/tango_sim_generator.py [hunk 10]
```diff
@@ -12,6 +12,7 @@
 from future import standard_library
 standard_library.install_aliases()  # noqa: E402
 from future.utils import with_metaclass
+from future.utils import itervalues
```
Comment (mmphego) [inline tango_simlib/tango_sim_generator.py]:
```suggestion
from future.utils import itervalues, with_metaclass
```

File: tango_simlib/tango_sim_generator.py [hunk 11]
```diff
@@ -188,7 +189,7 @@ def attr(tango_device_instance, new_val):
 
     # Sim test interface static attribute `attribute_name` info
     # Pick the first model instance in the dict.
-    controllable_attribute_names = models.values()[0].sim_quantities.keys()
+    controllable_attribute_names = list(itervalues(models))[0].sim_quantities.keys()
```
Comment (mmphego) [inline tango_simlib/tango_sim_generator.py]: IMHO this is complicating the not so complicated, how would you handle an exception.
This line of code would be intimidating for me to read if I was a noob.

Perhaps, break it down.

File: tango_simlib/tango_sim_generator.py [hunk 12]
```diff
@@ -224,7 +225,7 @@ def attr(tango_device_instance, new_val):
     # TODO(AR 02-03-2017): Ask the tango community on the upcoming Stack
     # Exchange community (AskTango) and also make follow ups on the next tango
     # releases.
-    for quantity_name, quantity in models.values()[0].sim_quantities.items():
+    for quantity_name, quantity in list(itervalues(models))[0].sim_quantities.items():
```
Comment (mmphego) [inline tango_simlib/tango_sim_generator.py]: Same here!

File: tango_simlib/tests/test_dish.py [hunk 13]
```diff
@@ -8,6 +8,9 @@
 from future import standard_library
 standard_library.install_aliases()  # noqa: E402
 
+from future.utils import itervalues
+
+
```
Comment (mmphego) [inline tango_simlib/tests/test_dish.py]: Run `Black` to fix this type of related style issues.

File: tango_simlib/tests/test_tango_sim_generator.py (hunk 3 above)
Comment (mmphego) [reply tango_simlib/tests/test_tango_sim_generator.py]: @lanceWilliams is correct, no need for list casting as `dict_value`'s are treated as lists when being iterated on.

image

File: tango_simlib/utilities/sim_xmi_parser.py [hunk 14]
```diff
@@ -420,9 +422,9 @@ def _get_arg_type(self, description_data):
 
         """
         if description_data.tag in ["attributes", "dynamicAttributes"]:
-            pogo_type = description_data.find("dataType").attrib.values()[0]
+            pogo_type = list(itervalues(description_data.find("dataType").attrib))[0]
```
Comment (mmphego) [inline tango_simlib/utilities/sim_xmi_parser.py]: "Simple is better than complex." - The Zen of Python, by Tim Peters

File: tango_simlib/utilities/sim_xmi_parser.py [hunk 15]
```diff
@@ -587,12 +589,12 @@ def get_device_attribute_metadata(self):
                             "{} information is not captured in the XMI"
                             " file".format(pogo_prop)
                         )
-            if isinstance(attribute_meta, unicode):  # noqa
-                attributes[attribute_meta["name"]] = attribute_meta.encode(
-                    "ascii", "replace"
-                )
-            else:
+
```
Comment (mmphego) [inline tango_simlib/utilities/sim_xmi_parser.py]: :+1:

File: tango_simlib/utilities/simdd_json_parser.py [hunk 16]
```diff
@@ -173,9 +173,10 @@ def get_device_data_components_dict(self, elements, element_type):
                 name = element_info["name"]
                 element_params = self._get_reformatted_data(element_info, element_type)
                 if "Attributes" in element_type:
-                    device_dict[str(name)] = dict(
-                        params_template.items() + element_params.items()
-                    )
+
```
Comment (mmphego) [inline tango_simlib/utilities/simdd_json_parser.py]: :+1:

File: tango_simlib/tests/test_tango_sim_generator.py (hunk 3 above)
Comment (tockards) [reply tango_simlib/tests/test_tango_sim_generator.py]: I thought so too but look at job 49 and job 50 on jenkins and notice the changes and the failed tests.

File: tango_simlib/compat.py (hunk 6 above)
Comment (tockards) [reply tango_simlib/compat.py]: Removed comment that was copied over from katcp compat.py

File: tango_simlib/utilities/sim_xmi_parser.py (hunk 14 above)
Comment (tockards) [reply tango_simlib/utilities/sim_xmi_parser.py]: do you have an alternative/suggestion of how to approach to this that will be simpler?

File: tango_simlib/tests/test_dish.py (hunk 13 above)
Comment (tockards) [reply tango_simlib/tests/test_dish.py]: ran `black` on all the files.

File: Jenkinsfile [hunk 17]
```diff
@@ -54,15 +54,20 @@ pipeline {
...
                 }
 
                 stage ('py36') {
                     steps {
-                         echo "Running nosetests on Python 3.6"
-                         sh 'python3 -m pip install -U .'
-                         sh "python3 setup.py nosetests --with-xunitmp --xunitmp-file=nosetests_py36.xml"
+                        echo "Running nosetests on Python 3.6"
+                        sh 'python3 -m pip install -U .'
+                        sh 'python3 -m coverage run --source="${KATPACKAGE}" -m nose --with-xunitmp --xunitmp-file=nosetests_py36.xml'
```
Comment (mmphego) [inline Jenkinsfile]: Confirmed `katcp` is baked into the image.

image

Comment (tockards) [reply Jenkinsfile]: thanks. It was not there when I was doing testing last week and hence the addition. It has been removed.

File: tango_simlib/compat.py (hunk 7 above)
Comment (tockards) [reply tango_simlib/compat.py]: These whitespaces was added by`black` should I remove it?

File: tango_simlib/tango_sim_generator.py (hunk 11 above)
Comment (tockards) [reply tango_simlib/tango_sim_generator.py]: I have changed it to multiple lines but what exceptions would you like me to handle?
Could that be done in a separate PR as `adding exception handling` seems to be outside of the scope of py2 to py3 conversion.

Comment (tockards) [reply tango_simlib/tango_sim_generator.py]: There are other alternatives. I opted for option 4 as it is the most explicit see https://python-future.org/compatible_idioms.html#dict-keys-values-items-as-a-list 

I could also bring in `dict` from `future.utils`  and just do a `list` casting.

File: tango_simlib/compat.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+
+from __future__ import absolute_import, division, print_function
+from future import standard_library
+
+standard_library.install_aliases()  # noqa: E402
+
+import future
+
+if future.utils.PY2:
+
```
[the remark this answers is not in the document]
Comment (tockards) [reply tango_simlib/compat.py]: This is an outcome of `black` perhaps @mmphego can give input on this?

**Comment by tockards** [comment]
clo
